### PR TITLE
feat(core): add waitUntil, a predicate wait scoped to the semantics tree

### PR DESCRIPTION
**Pull request by rock3r**
## Summary

Adds `ComposeAutomator.waitUntil(description, timeout, pollInterval) { … }`, the open-ended
counterpart to `waitForNode` / `waitUntilGone`, for barriers a tag/text selector cannot express: a
node count, a comparison, a combination of conditions, the shape of the tracked-window set. Refs #438.

```kotlin
automator.waitUntil(description = "the lazy list has realised at least five rows") {
    allNodes().count { it.testTag?.startsWith("row.") == true } >= 5
}
```

The predicate is an `AutomatorTree.() -> Boolean`, and that receiver **is** the scoping mechanism:
the semantics tree is the only thing handed to the condition, so the verb doesn't read as a
general-purpose "wait for anything" utility. Kotlin closures can still capture outside state, so
this is a boundary the API declines to invite rather than one it can enforce — the KDoc and the user
guide say exactly that, and say what to reach for instead (`withTimeout` around your own suspending
call, or an `AutomatorIdlingResource` if the state should also gate `waitForIdle`).

Each poll reads the tree on the EDT (`tree()` refreshes windows first, so a surface that appeared or
vanished between polls is observed rather than answered from a stale snapshot) and evaluates the
predicate on the calling coroutine, so a slow condition can't stall the UI it's watching. On timeout
it throws `IllegalStateException` naming the caller's `description` and the timeout, matching
`waitUntilGone`'s diagnostics style. `description` must be non-blank — it's the only thing the
failure can say about the condition — and that validation runs ahead of the EDT check, as in
`waitForNode` / `waitUntilGone`.

`AutomatorTree` deliberately gains **no** new members: `windows()`, `allNodes()`, and `roots()`
already express the conditions readably, so there was nothing to widen.

Three implementation notes worth a reviewer's attention:

- The deadline loop behind `waitUntilGoneInternal` is now a shared `pollUntilDeadline` with a
  caller-supplied timeout message — same timing contract (one observation minimum, one final
  observation at the deadline, sleeps clamped to the time left, injectable `clock`/`sleep` seam).
  `waitUntilGone`'s behaviour is unchanged and its existing tests still pin it, unmodified.
- The pre-existing internal top-level `waitUntil` (the `() -> T?` poller) is renamed
  `pollUntilNotNull` so one name in this package doesn't mean two different things. Resolution would
  have worked either way; the rename removes the trap.
- Adding a public verb pushed `ComposeAutomator` past detekt's `LargeClass` budget. Rather than
  suppress it, the UI-fingerprint formatting moved to a file-scope collaborator
  (`core/.../UiFingerprint.kt`) — a cohesive concern needing nothing from the automator but the
  tracked windows and the semantics reader. No `@Suppress`, no config change.

  ⚠️ **This was originally described as a pure move with no behaviour change, and that was wrong.**
  Both list hashes joined on a literal U+0001 separator; retyping the block dropped it, because a
  raw control byte renders as nothing in a terminal read *and* in the diff. Without it
  `["ab", "c"]` and `["a", "bc"]` hash identically, so re-segmented text would read as an unchanged
  fingerprint and let `waitForIdle` return on a UI that did change. Caught by Codex review, fixed in
  92bc3e4, and the separator is now a Kotlin escape rather than a raw byte so it can't go invisible
  again. The rest of the block was byte-compared against the original; those two separators were the
  only losses.

`core/api/core.api` is regenerated; the diff is two added lines and nothing removed.

Docs: new `waitUntil` section in `docs/guide/synchronization.md` (with a subsection on why it doesn't
wait on your program), its defaults in the table, and the "four wait helpers" phrasing updated to
five across the guide, `DOCS-STYLE.md`, and the skill files.

### Known limitation, deliberately not fixed here

`timeout` bounds the polling loop, not a single poll: each poll runs to completion before the
deadline is re-checked, so a blocking condition overruns it. This is family-wide rather than new —
`readOnEdt { tree() }` is an unbounded `invokeAndWait`, so a hung EDT overruns `waitUntilGone` and
`waitForNode` the same way. Making the timeout a true bound means running each poll on a worker with
the remaining budget and doing it across all the waits for consistency; that's a threading-contract
change for @rock3r to make deliberately, not something to slip into this PR. The KDoc and guide now
state the real contract instead of implying a stronger one.
[Discussion](https://github.com/rock3r/spectre/pull/489#discussion_r3916892930).

## Test plan

TDD per `docs/TESTING.md`: the tests went in first and were proven red (unresolved references) before
any production change, and the public-surface allowlist test failed exactly as expected before it was
updated in this same commit.

- [x] `./gradlew check` passes locally
- [x] `./gradlew ktfmtFormat` produced no changes

New coverage:

- `core/…/WaitUntilTest.kt` (11 tests) — public-boundary contracts against a headless automator
  (blank-description validation ahead of the EDT check, EDT rejection, a freshly-read tree per poll,
  the timeout message) plus the loop on virtual time through the internal seam, mirroring
  `WaitUntilGoneTest`.
- `core/…/UiFingerprintTest.kt` (4 tests) — pins the fingerprint's list-hash separator. This gap is
  why the regression above reached green CI: `WaitForIdleTest` stubs the fingerprint function
  wholesale, so nothing in the suite could observe how the fingerprint is *built*. Both failing
  cases were confirmed red against the separator-less version before the fix landed.
- `sample-desktop/…/WaitUntilValidationTest.kt` (4 tests) — live Compose: a node count, a secondary
  window entering and leaving the tracked set, a combined condition, and the timeout diagnostics.
  Registered in the required-class lists in both `validation-linux.yml` and `validation-windows.yml`.

Both suites were also run under `xvfb-run` locally, so the EDT-rejection tests actually executed
rather than being skipped as they are in a headless JVM, and all four live-Compose validation tests
passed against a real Compose surface. `mkdocs build --strict` is clean.

Two local failures were environmental and confirmed as such rather than assumed: `UdsPathLimitsTest`
fails identically on an unmodified `main` (this container's POSIX locale can't encode `é` in a path),
and `:cli:verifyCliDistributionZip` breaks because `JAVA_TOOL_OPTIONS` prints a banner as the first
line of `java -version`, which the launcher's `sed -n '1s/…/'` then parses. Both pass with
`LANG=C.UTF-8` and `JAVA_TOOL_OPTIONS` unset; neither is related to this change.

## Confirmations

- [ ] I read [CONTRIBUTING.md](../CONTRIBUTING.md) and this PR is **not** LLM-generated. (See the "no clanker PRs or issues" rule.)

  **Left unchecked deliberately — this PR is Claude-authored, at @rock3r's request in a Claude Code
  session.** Ticking it would be false. Flagging it here so the "no clanker PRs" rule is applied by
  you rather than bypassed by me.

- [x] I'm licensing this contribution under [Apache-2.0](../LICENSE).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Nhbq5cfftvk3ZjDv9GVWqz

### CODE REVIEW

File: core/src/main/kotlin/dev/sebastiano/spectre/core/UiFingerprint.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    if (node.isFocused) append(":F")
+    if (node.isDisabled) append(":D")
+    if (node.isSelected) append(":S")
+    if (node.texts.isNotEmpty()) {
+        append(":T")
+        append(node.texts.joinToString(separator = "").hashCode())
+    }
+    if (node.contentDescriptions.isNotEmpty()) {
+        append(":C")
+        append(node.contentDescriptions.joinToString(separator = "").hashCode())
```
[the remark this answers is not in the document]
Comment (rock3r) [reply core/src/main/kotlin/dev/sebastiano/spectre/core/UiFingerprint.kt]: Correct, and this was a real regression — fixed in 92bc3e4.

Verified against the bytes rather than my own reading of the diff: `git show HEAD~1:…/ComposeAutomator.kt | grep joinToString | cat -A` shows `separator = "^A"` on both lines. The separator was a raw U+0001 control byte, which renders as nothing in a terminal read *and* in the diff — I retyped that block by hand during the extraction and dropped it without either of us seeing it. My "pure extraction, no behaviour change" claim in the commit and PR body was wrong; the PR body is corrected.

I also byte-compared the whole original fingerprint block against the extracted one to check nothing else went missing the same way. Those two separators were the only losses.

Two changes rather than one:

- The separator is restored as the Kotlin escape `""` instead of a raw control byte, so it stays visible to the next person who reads, copies, or reviews the line. Same string value, so the fingerprint is byte-identical to pre-extraction.
- `fingerprintHashOf` is now a named seam with `UiFingerprintTest` pinning it. This bug reached CI green because `WaitForIdleTest` stubs the fingerprint function wholesale — nothing in the suite could observe how the fingerprint is *built*. Both cases you named (`["ab","c"]` vs `["a","bc"]`, and the element-boundary case) were confirmed failing against the separator-less version before the fix went in, so the test demonstrably catches this exact bug rather than just passing.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: core/src/main/kotlin/dev/sebastiano/spectre/core/ComposeAutomator.kt [hunk 2]
```diff
@@ -911,6 +870,49 @@ private constructor(
...
+    ) {
+        // Same ordering as waitForNode / waitUntilGone: bad arguments are the more actionable
+        // error for an EDT caller doing two things wrong.
+        require(description.isNotBlank()) { "description must not be blank" }
+        rejectEdtCaller("waitUntil")
+        waitUntilInternal(
+            timeout = timeout,
+            pollInterval = pollInterval,
+            description = description,
+            condition = { readOnEdt { tree() }.condition() },
```
[the remark this answers is not in the document]
Comment (rock3r) [reply core/src/main/kotlin/dev/sebastiano/spectre/core/ComposeAutomator.kt:922]: The observation is right — `pollUntilDeadline` calls the predicate synchronously and re-checks the clock only after it returns, so `waitUntil(timeout = 100.milliseconds) { Thread.sleep(10_000); false }` does block for ~10s. Fixed in 92bc3e4, but by correcting the contract rather than by bounding the lambda, and that's worth explaining.

Bounding only the predicate would fix the least likely half of the problem. The poll's other half is `readOnEdt { tree() }` — a `SwingUtilities.invokeAndWait` that is itself unbounded, so a hung EDT overruns the timeout even when the caller's condition is trivial. That is true of `waitUntilGone` (whose `countPresent` does the same EDT round-trip) and of `waitForNode` too, since `withTimeout` cannot cancel a blocking non-suspending call either. So the family has never offered "the timeout bounds the whole wait"; making `waitUntil` alone appear to would advertise a guarantee its siblings don't keep, on the half that fails least often.

Making it a true bound means running each poll on a worker with the remaining budget — the `runBoundedOnWorker` pattern `waitForIdle` already uses for fingerprint sampling — and applying it across all four existing waits for consistency. That is a deliberate change to the threading contract of the whole wait family, not something to slip into a PR adding one verb, so it should be @rock3r's call in its own change rather than mine here.

What this PR does instead is stop the KDoc and the user guide overclaiming: `timeout` bounds the polling loop, not a single poll; each poll runs to completion before the deadline is re-checked; keep the condition cheap and side-effect free. That last point is the same discipline the `AutomatorTree`-only receiver is aimed at, arrived at from the other direction — a condition that only reads the tree has nothing to block on.

Leaving this thread open rather than resolving it, since the underlying limitation is real and unfixed.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by rock3r** [comment]
### `validation-linux` failed on `92bc3e4` — not this PR's, re-running once

**What failed:** step 8, `Run real-Robot Linux smoke under Xvfb` (`:sample-desktop:runLinuxRobotSmoke`), scoring **0/7**:

```
mouseMove + click on counter: FAIL — clickCount 0 → 0 at (270,96)
doubleClick on counter:       FAIL — clickCount 0 → 0
typeText into BasicTextField: FAIL — text="" expected="hello linux"
pressKey 'h':                 FAIL — before="" after=""
clearAndTypeText:             FAIL — text="" expected="replaced"
Ctrl+S via pressKey:          FAIL — shortcutFiredCount 0 → 0
screenshot of color patch:    FAIL — sampled center pixel is solid black (#000000)
--- 0/7 passed ---
```

Step 7 — `Run validation tests under Xvfb`, which is where this PR's new `WaitUntilValidationTest` runs — **passed**, and the JUnit verifier reported "All required classes present; 23 discovered XML report(s) clean."

**Why it isn't this PR's:**

1. `LinuxRobotSmoke.kt` and `RobotSmokeRig.kt` contain no reference to `waitForIdle`, `waitForVisualIdle`, `fingerprint`, or `waitUntil`. Nothing this PR changes is on that code path — the diff is a wait verb, a fingerprint-formatting extraction, tests, and docs.
2. This exact job **passed on `c9d3bee`**, this PR's previous commit ([run 33661607894](https://github.com/rock3r/spectre/actions/runs/33661607894)). The only delta to `92bc3e4` is the fingerprint separator restore, a new `:core` unit test, and prose — none of it reachable from a Robot smoke.
3. `main` is green on this workflow (`6026ba29`, [run at 16:23Z](https://github.com/rock3r/spectre/actions/runs/33662714024)).
4. `runLinuxRobotUnfocusedSmoke` ran on the **same runner seconds later and passed 4/4**, including `click on unfocused counter increments + transfers focus — AWT mouse press dispatched in this JVM`. The Robot input stack demonstrably worked on that display.

**What it looks like instead:** every scenario failed, including the screenshot sampling solid black — which is the symptom the smoke's own message calls out as "the JFrame was occluded or off-screen when the capture fired". All seven failing together, with the unfocused variant fine immediately after, points at the SUT window never being mapped/composited on the Xvfb display for that run, not at a logic change.

I'm not touching the smoke or this PR's code for it: no fix exists to port, and skipping or weakening the smoke to get green isn't on the table. Re-running the failed job once to confirm — that's the single re-run spent. If it reproduces identically, it stops being a flake in my book and I'll come back with a root-cause investigation rather than another re-run.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by rock3r** [comment]
### Re-run did **not** reproduce — it failed differently

Correcting my previous comment: I said an identical reproduction would stop me calling this a flake. It didn't reproduce identically, so that condition wasn't met.

| | attempt 1 (runner …170) | attempt 2 (runner …205) |
|---|---|---|
| Score | **0/7** | **6/7** |
| Frame bounds | `x=0,y=0,540x320` | `x=370,y=352,540x320` |
| Failures | all seven; screenshot **solid black** | only `clearAndTypeText` |
| Screenshot | FAIL (`#000000`) | PASS (`#3366CC`, exact match) |

Attempt 2's single failure:

```
clearAndTypeText: FAIL — text="hello linuxreplaced" expected="replaced"
```

The clear didn't take — the prior scenario's `"hello linux"` survived and `"replaced"` was appended. Clicks, double-clicks, `typeText`, `pressKey`, Ctrl+S and the screenshot all passed in that same run, so real input and compositing were working; this is a select-all/clear timing miss, not the dead-display symptom of attempt 1.

Two unrelated failure modes across two runners is flakiness in a timing-sensitive real-input smoke, not a deterministic breakage.

**Also worth pinning down: the smoke never touches this PR's code.** `RobotSmokeRig` drives `RobotDriver` directly — `driver.clearAndTypeText(...)` is `RobotDriver.clearAndTypeText`, not `ComposeAutomator.clearAndTypeText`. There is no `ComposeAutomator`, `waitForIdle`, `waitUntil` or fingerprint usage anywhere in `RobotSmokeRig.kt` or `LinuxRobotSmoke.kt`, and this PR touches no line of `RobotDriver`.

**One hypothesis I can't rule out, in fairness.** This PR adds a validation class to step 7, making that step longer (2m15s → 3m04s across the two attempts) and leaving more JVMs behind — the job's cleanup terminates several orphaned `java` processes. Each step gets its own `xvfb-run -a` display, so cross-step occlusion isn't the mechanism, but extra background CPU contention could plausibly make an already timing-sensitive smoke flake more often. That would make this PR a *contributor* to flake frequency rather than a cause of breakage. Testing that properly means investigating the smoke's timing margins, which is its own change, not this one.

My one re-run is spent and I'm not spending another. Not skipping, quarantining, or weakening the smoke to get green, and not widening this PR to retune unrelated input-timing code. Over to @rock3r on how to proceed.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by rock3r** [comment]
### Reproduced locally: the `clearAndTypeText` failure is pre-existing, and my last comment's hypothesis was wrong

Ran the smoke under Xvfb in a Linux container. Correcting two things I said earlier.

**1. It reproduces with no validation step in front of it.** My hypothesis that step 7 was starving step 8 of CPU is dead — running `:sample-desktop:runLinuxRobotSmoke` on its own reproduces the identical failure (`text="hello linuxreplaced"`, 6/7). No step 7 involved.

**2. It reproduces on this PR's merge base, which contains none of this PR's code.** `6026ba29` fails the same way.

**3. Paired A/B, base vs head, interleaved to control for machine load:**

| round | base `6026ba29` | head `92bc3e4` |
|---|---|---|
| 1 | 6/7 | 7/7 |
| 2 | 6/7 | 6/7 |
| 3 | 7/7 | 6/7 |
| 4 | 6/7 | 6/7 |
| **failures** | **3 / 4** | **3 / 4** |

Identical rates. I'd earlier measured 4/5 on head against 1/4 on base and taken it seriously — that was an artifact of running the two batches at different times under different container load. Interleaving removes it and the difference disappears. (Two git worktrees, so neither side pays a recompile the other doesn't.)

**Mechanism**, from `RobotDriver.clearAndTypeText`:

```kotlin
robot.keyPress(selectAllModifier); robot.keyPress(VK_A); robot.keyRelease(VK_A)
robot.keyRelease(selectAllModifier)
robot.keyPress(VK_BACK_SPACE); robot.keyRelease(VK_BACK_SPACE)
typeText(text)
```

No wait or verification between the select-all and the backspace. When Compose hasn't processed Ctrl+A by the time Backspace lands, the backspace deletes a single character at the caret instead of the selection — which is exactly the observed `"hello linux"` + `"replaced"`, with only the trailing `h` gone. This PR touches no line of `RobotDriver`, and `RobotSmokeRig` drives it directly without ever constructing a `ComposeAutomator`.

This looks worth fixing on its own — the smoke is asserting real behaviour and the race is in production input code, not the test — but retuning input timing is not this PR's business, so I've left it alone rather than widening the diff. Happy to open a separate issue or PR for it if you want.

### Also pushed: `main` merged in, plus one docs fix it made necessary

The branch was based on `6026ba29`, so I brought `main` (`fb0c25b`) in. One conflict in `docs/guide/synchronization.md` — both sides added text at the same seam, no semantic clash; kept both, with #490's transport-parity paragraph closing the `waitUntilGone` section and the `waitUntil` section after it.

That merge surfaced a genuine accuracy problem. #490 documents `waitUntilGone` reaching every transport `waitForNode` does, and that paragraph now sits directly above the new `waitUntil` section — so a reader would assume the same reach. `waitUntil` doesn't have it and can't: the condition is a Kotlin lambda, and a lambda doesn't cross a process boundary, so there's nothing to route over CLI, MCP, or agent attach. The KDoc and guide now say so and point remote callers at a selector wait.

`./gradlew check` passes on the merged tree, including #490's contract corpus and capability-matrix tests — nothing there requires the new verb to be routed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: core/src/main/kotlin/dev/sebastiano/spectre/core/WaitUtilities.kt [hunk 3]
```diff
@@ -23,46 +23,104 @@ internal suspend fun <T : Any> waitUntil(
...
+    satisfied: () -> Boolean,
+    timeoutMessage: () -> String,
+    clock: MonotonicClock,
+    sleep: suspend (Duration) -> Unit,
 ) {
     val deadline = clock.now() + timeout.inWholeMilliseconds
     while (true) {
-        val present = countPresent()
-        if (present == 0) return
+        if (satisfied()) return
```
[the remark this answers is not in the document]
Comment (rock3r) [reply core/src/main/kotlin/dev/sebastiano/spectre/core/WaitUtilities.kt:59]: Mechanically correct — the post-sleep `satisfied()` is accepted without comparing the new time to `deadline`, so a wait can return success past its advertised timeout. Addressed in 54885da, but by pinning the behaviour rather than changing it, and the reason matters.

**This isn't new here.** The loop is verbatim the one `waitUntilGone` shipped with in #482:

```kotlin
// 6e39c5d, waitUntilGoneInternal
while (true) {
    val present = countPresent()
    if (present == 0) return              // accepted regardless of the clock
    val remainingMs = deadline - clock.now()
    if (remainingMs <= 0) { error(...) }
    sleep(minOf(pollInterval, remainingMs.milliseconds))
}
```

Extracting `pollUntilDeadline` preserved that structure exactly — which was the point, since this PR claims `waitUntilGone`'s behaviour is unchanged. So the strict fix isn't a fix to new code; it's a semantics change to a shipped verb, and it would land in a PR whose stated contract is that it changes nothing about `waitUntilGone`.

**And I think the current behaviour is the right one.** Rejecting an observation that found the condition true means throwing `condition "…" never held in tracked windows` immediately after watching it hold — a diagnostic that is simply false. The alternative you describe (check the clock first, don't observe) throws without looking, which is less provably wrong but no more informative. Either way a wait whose UI *did* settle, a few milliseconds past the deadline on a loaded machine, becomes a failure. For a UI test-automation library that manufactures exactly the flakiness this repo already spends CI budget fighting — this very PR is sitting behind a `validation-linux` red caused by a millisecond-scale input race.

What was genuinely missing is that none of this was written down or tested, so it read as an oversight rather than a decision. 54885da adds a characterisation test driving `sleep` to overshoot the deadline and asserting the wait still returns, plus a KDoc paragraph stating that the timeout is not a hard bound on returning success and why.

Leaving this thread open: the strict-deadline option is a real alternative, and if @rock3r prefers it, it should be applied to `waitUntilGone` and `waitUntil` together in its own change, with the timeout-message wording revisited so it doesn't assert something false.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by rock3r** [comment]
### `check-windows` went red on `0bb5f27` — inherited from `main`, not from this PR

Two test failures, both IPC/socket timing, both in modules this PR does not touch:

```
:cli:test    DaemonServerTest > idle cleanup keeps an accepted client alive through its first request()
             AssertionFailedError at DaemonServerTest.kt:33
:agent:test  LongOpInfrastructureTest > ordinary EOF accepts a replacement client while input finishes()
             AssertionFailedError at LongOpInfrastructureTest.kt:106
```

**`main` is red on `windows.yml` with the same failure.** Run [33666304345](https://github.com/rock3r/spectre/actions/runs/33666304345) on `fb0c25b` fails on `LongOpInfrastructureTest > ordinary EOF accepts a replacement client while input finishes()` at the same line 106. The preceding main commit `bc40bf08` was green, so this arrived with `fb0c25b` (#490).

This PR went red on it because I merged `main` in (`26b19f4`) to get off a stale base — so it inherited main's breakage along with #490's transport work. `check-windows` was green here on `92bc3e4`, the last head before that merge. This PR's own diff is `:core` plus tests and docs; it touches no line of `:cli` or `:agent`.

**No fix to port yet** — `main`'s tip is the red commit, so there is nothing green to bring across. When one lands, merging `main` again picks it up and no-ops here.

I'm not re-running: my one re-run was spent on `validation-linux`, and `54885da` is getting a fresh natural CI run anyway (in flight now).

Worth flagging separately for whoever owns #490: `DaemonServerTest.kt:33` failed here but not in main's run, so between the two there are two distinct daemon/IPC socket-timing tests going red on Windows since that commit — that looks like one underlying race rather than two coincidences.

---
_Generated by [Claude Code](https://claude.ai/code)_
